File: leetcode/ruby/113_path_sum_II.rb
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

def path_sum(root, sum)
  return [] unless root
  res = []
  path_sum_helper(root, sum, [], res)
  res
end

def path_sum_helper(root, sum, route, res)
  if root.left.nil? && root.right.nil?
    if root.val == sum
      route += [root.val]
      res << route
    end
    # early return because there is no left or right to call.
    return res
  end

  path_sum_helper(root.left, sum - root.val, route + [root.val], res) if root.left
  path_sum_helper(root.right, sum - root.val, route + [root.val], res) if root.right
end

File: leetcode/python/113_path_sum_ii.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        res = []
        self.dfs(root, sum, [], res)
        return res

    def dfs(self, root: TreeNode, sum: int, route: List[int], res: List[List[int]]):
        if not root.left and not root.right and sum == root.val:
            route.append(root.val)
            res.append(route)
        if root.left:
            self.dfs(root.left, sum-root.val, route + [root.val], res)
        if root.right:
            self.dfs(root.right, sum-root.val, route + [root.val], res)
